### ISSUE TITLE
problem with concat in content_tag block

### ISSUE THREAD

**Issue by samstickland** [apotonick/cells#294]
Hi,

I'm having a issue with the use of concat in content_tag block contents. In bootstrap_form_for the following code is called:

          content_tag :div, class: css do
            concat content_tag :p, title
            concat error_summary unless options[:error_summary] == false
          end

This generates a div tag, but it doesn't have any content. I've tried stepping through the code, but I can't find the problem.

I can see that both content_tag and concat are delegated to @template, and the OutputBuffer that gets created is derived from Array (in Cells), and that concat gets mapped to << which in turns calls super. And I can see that the array gets the contents of title in a p tag. But this doesn't seem to get end up in the div tag.

Any ideas?

**Comment by apotonick** [comment]
Cells 3 or 4?

**Comment by samstickland** [comment]
Cells 4.0.0

**Comment by samstickland** [comment]
Right, I've tracked this to the with_output_buffer method called from capture

When the ActionView method invokes yield it captures the block output in output_buffer:

```
[196, 205] in /home/dev/.rvm/gems/ruby-2.2.1/gems/actionview-4.2.1/lib/action_view/helpers/capture_helper.rb
   196:           buf = ActionView::OutputBuffer.new
   197:           buf.force_encoding(output_buffer.encoding) if output_buffer
   198:         end
   199:         self.output_buffer, old_buffer = buf, output_buffer
   200:         yield
=> 201:         output_buffer
   202:       ensure
   203:         self.output_buffer = old_buffer
   204:       end
   205:     end
(byebug) p output_buffer
"<p>Please fix the errors below.</p>"
```

But the Cells implementation isn't capturing the output in the returned variable.

```
[33, 42] in /home/dev/.rvm/gems/ruby-2.2.1/bundler/gems/cells-erb-1ac42b5432bf/lib/cell/erb/template.rb
   33:     def with_output_buffer(block_buffer=ViewModel::OutputBuffer.new)
   34:       @output_buffer, old_buffer = block_buffer, @output_buffer
   35:       yield
   36:       @output_buffer = old_buffer
   37: 
=> 38:       block_buffer
   39:     end
   40: 
   41:     # Below:
   42:     # Rails specific helper fixes. I hate that. I can't tell you how much I hate those helpers,
(byebug) p block_buffer
[]
```

But I don't understand how either of the buffers are supposed to magically capture the output of yield so I don't know how to go about fixing this :)

**Comment by samstickland** [comment]
cells and cells-erb are both at master btw.

**Comment by samstickland** [comment]
OK, so I figured out that it's the concat method that should add to the output_buffer and it's the concat method that, bizarrely, doesn't seem to be working when output_buffer is Cell::ViewModel::OutputBuffer.

Here's some byebug output inside the concat method where I try to manually run the << method to append to the Cell::ViewModel::OutputBuffer object without success.

```
[48, 57] in /home/dev/.rvm/gems/ruby-2.2.1/gems/actionview-4.2.1/lib/action_view/helpers/text_helper.rb
   48:       #         concat link_to('login', action: :login)
   49:       #       end
   50:       #       # will either display "Logged in!" or a login link
   51:       #   %>
   52:       def concat(string)
=> 53:         output_buffer << string
   54:       end
   55: 
   56:       def safe_concat(string)
   57:         output_buffer.respond_to?(:safe_concat) ? output_buffer.safe_concat(string) : concat(string)
(byebug) p output_buffer
[]
[]
(byebug) p string
"<p>Please fix the errors below.</p>"
"<p>Please fix the errors below.</p>"
(byebug) output_buffer << string
["<p>Please fix the errors below.</p>"]
(byebug) output_buffer.join
""
(byebug) output_buffer << "<p>Please fix the errors below.</p>"
["<p>Please fix the errors below.</p>"]
(byebug) output_buffer.join
""
```

It works just fine in a rails console though :/

```
2.2.1 :008 > output_buffer = Cell::ViewModel::OutputBuffer.new
 => [] 
2.2.1 :009 > output_buffer << "<p>Please fix the errors below.</p>"
 => ["<p>Please fix the errors below.</p>"] 
2.2.1 :010 > output_buffer.join
 => "<p>Please fix the errors below.</p>" 
```

**Comment by apotonick** [comment]
Do you get a feeling now why I hate Rails helpers and don't want to do anything with them anymore?

BTW I know where the problem is, stop wasting your precious time.. :stuck_out_tongue_winking_eye:

**Comment by samstickland** [comment]
Well it's all learning! I'm interested to know what the fix is :)

2015-06-22 22:47 GMT+01:00 Nick Sutterer <notifications@github.com>:

**Comment by apotonick** [comment]
This is not learning because the way helpers are done and output buffering is done in Rails is _WRONG_. It took me some years to realise that I should better not learn from Rails.

**Comment by apotonick** [comment]
Fixed in cells-erb 0.0.5.

The problem was that the old helper uses `output_buffer` (the method). In cells, this method is only called once per view, so this creates a new output buffer and thus forgets some input (as we're appending to two different output buffers).

If you want to learn how it's done right, check out how slim does it. No need to fix anything there, as it's really clean. I will soon change Erbse to support these operations on the ERB level.

In case you're interested in all this and want to help - just let me know, I can give you an introduction and then maybe you want to play with Erbse a bit? :cake:

**Comment by samstickland** [comment]
Ah, I see, that makes sense. I've upgraded and it's all working for me now.

With regards, the help, I'm a bit busy at the moment (just finishing up one
network consulting job and doing some software development on the side),
but once things settle down a but I think I might be interested :)

S

2015-06-23 0:50 GMT+01:00 Nick Sutterer <notifications@github.com>:
